File: src/main/java/offkilter/infohud/InfoHUD.java
package offkilter.infohud;

import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents;
import net.fabricmc.fabric.api.networking.v1.*;
import net.minecraft.Util;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.server.network.ServerGamePacketListenerImpl;
import net.minecraft.util.Mth;
import net.minecraft.world.level.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import java.util.*;

public class InfoHUD implements ModInitializer {
    private long lastTickPerfSend = 0;
    private long lastServerLightSend = 0;
    private int tickCount = 0;
    public final Logger logger = LogManager.getLogger("infohud");

    private final Set<UUID> tickPerfClients = new HashSet<>();
    private final Set<UUID> serverLightClients = new HashSet<>();

    @Override
    public void onInitialize() {
        ServerTickEvents.END_SERVER_TICK.register(this::serverTickEnd);
        S2CPlayChannelEvents.REGISTER.register((handler, sender, server, channels) -> handleRegistrationAdded(handler, channels));
        S2CPlayChannelEvents.UNREGISTER.register((handler, sender, server, channels) -> handleRegistrationRemoved(handler, channels));
        ServerPlayConnectionEvents.DISCONNECT.register(this::playerDisconnected);
    }

    private void handleRegistrationAdded(ServerGamePacketListenerImpl handler, List<ResourceLocation> channels) {
        if (channels.contains(InfoHUDNetworking.TICK_PERF)) {
            long oldCount = tickPerfClients.size();
            tickPerfClients.add(handler.getPlayer().getUUID());
            if (oldCount == 0) {
                logger.info("Tick perf now enabled");
            }
        }
        if (channels.contains(InfoHUDNetworking.SERVER_LIGHT)) {
            long oldCount = serverLightClients.size();
            serverLightClients.add(handler.getPlayer().getUUID());
            if (oldCount == 0) {
                logger.info("Server light now enabled");
            }
        }
    }

    private void handleRegistrationRemoved(ServerGamePacketListenerImpl handler, List<ResourceLocation> channels) {
        if (channels.contains(InfoHUDNetworking.TICK_PERF)) {
            tickPerfClients.remove(handler.getPlayer().getUUID());
            if (tickPerfClients.size() == 0) {
                logger.info("Tick perf now disabled");
            }
        }
        if (channels.contains(InfoHUDNetworking.SERVER_LIGHT)) {
            serverLightClients.remove(handler.getPlayer().getUUID());
            if (serverLightClients.size() == 0) {
                logger.info("Server light now disabled");
            }
        }
    }

    private void playerDisconnected(ServerGamePacketListenerImpl handler, MinecraftServer server) {
        if (tickPerfClients.remove(handler.getPlayer().getUUID()) && (tickPerfClients.size() == 0)) {
            logger.info("Tick perf now disabled");
        }
        if (serverLightClients.remove(handler.getPlayer().getUUID()) && (serverLightClients.size() == 0)) {
            logger.info("Server light now disabled");
        }
    }

    private void serverTickEnd(@NotNull MinecraftServer server) {
        if (!tickPerfClients.isEmpty()) {
            tickCount += 1;
            long now = Util.getMillis();
            if ((now - lastTickPerfSend) >= 1000) {
                FriendlyByteBuf buf = PacketByteBufs.create();
                buf.writeLong((long) Mth.average(server.tickTimes));
                buf.writeLong(tickCount);

                for (ServerPlayer p : server.getPlayerList().getPlayers()) {
                    if (ServerPlayNetworking.canSend(p, InfoHUDNetworking.TICK_PERF)) {
                        ServerPlayNetworking.send(p, InfoHUDNetworking.TICK_PERF, buf);
                    }
                }

                lastTickPerfSend = now;
                tickCount = 0;
            }
        }

        if (!serverLightClients.isEmpty()) {
            long now = Util.getMillis();
            if ((now - lastServerLightSend) >= 500) {
                ServerLevel level = server.getLevel(Level.OVERWORLD);
                if (level != null) {
                    FriendlyByteBuf buf = PacketByteBufs.create();
                    buf.writeInt(level.isThundering() ? 10 : level.getSkyDarken());
                    for (ServerPlayer p : server.getPlayerList().getPlayers()) {
                        if (ServerPlayNetworking.canSend(p, InfoHUDNetworking.SERVER_LIGHT)) {
                            ServerPlayNetworking.send(p, InfoHUDNetworking.SERVER_LIGHT, buf);
                        }
                    }
                }
                lastServerLightSend = now;
            }
        }
    }
}

File: src/main/kotlin/offkilter/infohud/InfoHUD.kt
package offkilter.infohud

import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents
import net.fabricmc.fabric.api.networking.v1.*
import net.minecraft.Util
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.MinecraftServer
import net.minecraft.server.network.ServerGamePacketListenerImpl
import net.minecraft.util.Mth
import net.minecraft.world.level.Level
import org.apache.logging.log4j.LogManager
import java.util.*

class InfoHUD : ModInitializer {
    private var lastTickPerfSend: Long = 0
    private var lastServerLightSend: Long = 0
    private var tickCount = 0
    val logger = LogManager.getLogger("infohud")
    private val tickPerfClients: MutableSet<UUID> = HashSet()
    private val serverLightClients: MutableSet<UUID> = HashSet()
    override fun onInitialize() {
        ServerTickEvents.END_SERVER_TICK.register(ServerTickEvents.EndTick { server: MinecraftServer ->
            serverTickEnd(
                server
            )
        })
        S2CPlayChannelEvents.REGISTER.register(S2CPlayChannelEvents.Register { handler: ServerGamePacketListenerImpl, sender: PacketSender?, server: MinecraftServer?, channels: List<ResourceLocation> ->
            handleRegistrationAdded(
                handler,
                channels
            )
        })
        S2CPlayChannelEvents.UNREGISTER.register(S2CPlayChannelEvents.Unregister { handler: ServerGamePacketListenerImpl, sender: PacketSender?, server: MinecraftServer?, channels: List<ResourceLocation> ->
            handleRegistrationRemoved(
                handler,
                channels
            )
        })
        ServerPlayConnectionEvents.DISCONNECT.register(ServerPlayConnectionEvents.Disconnect { handler: ServerGamePacketListenerImpl, server: MinecraftServer ->
            playerDisconnected(
                handler,
                server
            )
        })
    }

    private fun handleRegistrationAdded(handler: ServerGamePacketListenerImpl, channels: List<ResourceLocation>) {
        if (channels.contains(InfoHUDNetworking.TICK_PERF)) {
            val oldCount = tickPerfClients.size.toLong()
            tickPerfClients.add(handler.getPlayer().uuid)
            if (oldCount == 0L) {
                logger.info("Tick perf now enabled")
            }
        }
        if (channels.contains(InfoHUDNetworking.SERVER_LIGHT)) {
            val oldCount = serverLightClients.size.toLong()
            serverLightClients.add(handler.getPlayer().uuid)
            if (oldCount == 0L) {
                logger.info("Server light now enabled")
            }
        }
    }

    private fun handleRegistrationRemoved(handler: ServerGamePacketListenerImpl, channels: List<ResourceLocation>) {
        if (channels.contains(InfoHUDNetworking.TICK_PERF)) {
            tickPerfClients.remove(handler.getPlayer().uuid)
            if (tickPerfClients.size == 0) {
                logger.info("Tick perf now disabled")
            }
        }
        if (channels.contains(InfoHUDNetworking.SERVER_LIGHT)) {
            serverLightClients.remove(handler.getPlayer().uuid)
            if (serverLightClients.size == 0) {
                logger.info("Server light now disabled")
            }
        }
    }

    private fun playerDisconnected(handler: ServerGamePacketListenerImpl, server: MinecraftServer) {
        if (tickPerfClients.remove(handler.getPlayer().uuid) && tickPerfClients.size == 0) {
            logger.info("Tick perf now disabled")
        }
        if (serverLightClients.remove(handler.getPlayer().uuid) && serverLightClients.size == 0) {
            logger.info("Server light now disabled")
        }
    }

    private fun serverTickEnd(server: MinecraftServer) {
        if (!tickPerfClients.isEmpty()) {
            tickCount += 1
            val now = Util.getMillis()
            if (now - lastTickPerfSend >= 1000) {
                val buf = PacketByteBufs.create()
                buf.writeLong(server.tickTimes.toList().average().toLong())
                buf.writeLong(tickCount.toLong())
                for (p in server.playerList.players) {
                    if (ServerPlayNetworking.canSend(p, InfoHUDNetworking.TICK_PERF)) {
                        ServerPlayNetworking.send(p, InfoHUDNetworking.TICK_PERF, buf)
                    }
                }
                lastTickPerfSend = now
                tickCount = 0
            }
        }
        if (!serverLightClients.isEmpty()) {
            val now = Util.getMillis()
            if (now - lastServerLightSend >= 500) {
                val level = server.getLevel(Level.OVERWORLD)
                if (level != null) {
                    val buf = PacketByteBufs.create()
                    buf.writeInt(if (level.isThundering) 10 else level.skyDarken)
                    for (p in server.playerList.players) {
                        if (ServerPlayNetworking.canSend(p, InfoHUDNetworking.SERVER_LIGHT)) {
                            ServerPlayNetworking.send(p, InfoHUDNetworking.SERVER_LIGHT, buf)
                        }
                    }
                }
                lastServerLightSend = now
            }
        }
    }
}